### ISSUE TITLE
Synchronization at the end of DASH algorithms

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#376]
While working on #374 I realized that some algorithms in DASH synchronize units at the end (e.g., `dash::for_each`) and some don't (e.g., `dash::generate`). This should be unified. 

@fuchsto I assume this is covered by Execution Policies (#104)? If so, this is just to track progress on this issue.

**Comment by fuchsto** [comment]
Yes, exactly, but it's sensible to track this as a separate issue.
